shotover-proxy/src/transforms/redis/sink_cluster.rs

```diff
@@ -378,66 +381,85 @@ impl RedisSinkCluster {
                 Ok(Err(e)) => {
                     debug!("failed to connect to {}: {}", host, e);
                     self.rebuild_connections = true;
-                    self.short_circuit(one_tx);
-                    return Ok(one_rx);
+                    return self.short_circuit_with_error();
                 }
                 Err(_) => {
                     debug!("timed out connecting to {}", host);
                     self.rebuild_connections = true;
-                    self.short_circuit(one_tx);
-                    return Ok(one_rx);
+                    return self.short_circuit_with_error();
                 }
             }
         };
 
-        if let Err(e) = channel.send(Request {
-            message,
-            return_chan: Some(one_tx),
-        }) {
-            if let Some(error_return) = e.0.return_chan {
-                self.rebuild_connections = true;
-                self.short_circuit(error_return);
-            }
+        let (one_tx, one_rx) = oneshot::channel::<Response>();
+        if channel
+            .send(Request {
+                message,
+                return_chan: Some(one_tx),
+            })
+            .is_err()
+        {
+            self.rebuild_connections = true;
             self.channels.remove(host);
+            return self.short_circuit_with_error();
         }
 
-        Ok(one_rx)
+        Ok(Box::pin(one_rx.map_err(|e| anyhow!(e))))
     }
 
-    #[inline(always)]
-    fn get_channels(&mut self, routing_info: RoutingInfo) -> ChannelResult {
+    async fn dispatch_message_hiding(
+        &mut self,
+        routing_info: RoutingInfo,
+        message: Message,
+    ) -> Result<ResponseFuture> {
         match routing_info {
-            RoutingInfo::Slot(slot) => ChannelResult::Channels(
-                if let Some((_, lookup)) = self.slots.masters.range(&slot..).next() {
-                    vec![lookup.clone()]
-                } else {
-                    vec![]
-                },
-            ),
+            RoutingInfo::Slot(slot) => self.send_message_to_slot(slot, message).await,
             RoutingInfo::AllNodes(_) => {
-                ChannelResult::Channels(self.slots.nodes.iter().cloned().collect())
+                self.send_message_to_channels(
+                    &self.slots.nodes.iter().cloned().collect_vec(),
```

Comment:
Is it possible to not clone here?

Reply:
It could be possible with a bit more fiddling and would be a nice win but this was as far as I got before leaving it as is.
![image](https://user-images.githubusercontent.com/5120858/170428132-383cd960-4ee5-4d82-8173-338e73d6fa87.png)